### ISSUE TITLE
Syslogd cannot be started when container is started with --read-only

### ISSUE THREAD

**Issue by piotrminkina** [moby/moby#15938]
When container is started with option `--read-only` then `syslogd` cannot be started, because `/dev` is also read only and process cannot create socket at `/dev/log`. In previous versions of Docker was not any problems.

Could You add possibility of select that `/dev` or `/sys` should also be read only when option `--read-only` will be provided? Maybe You have better idea to solve this?

```
# uname -a
Linux gaia 4.1.0-2-amd64 #1 SMP Debian 4.1.6-1 (2015-08-23) x86_64 GNU/Linux
```

```
# docker version
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:24:20 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:24:20 UTC 2015
 OS/Arch:      linux/amd64
```

```
# docker -D info
Containers: 9
Images: 146
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.0-2-amd64
Operating System: Debian GNU/Linux stretch/sid
CPUs: 4
Total Memory: 11.63 GiB
Name: gaia
ID: B7WF:V6UQ:DRVY:CL72:HGTY:DA7L:NVC7:XJ47:I3IG:3ZD3:UZX7:DWY2
Username: piotrminkina
Registry: https://index.docker.io/v1/
```

Regards and thanks for great product!
Piotr Minkina

**Comment by thaJeztah** [comment]
Just to check, this is about running syslogd *inside* a container (not using the syslog logging driver; https://docs.docker.com/reference/run/#logging-drivers-log-driver)?

**Comment by piotrminkina** [comment]
I know about this functionality, but this works outside container, not internal. Some services, like Postfix MTA can only log to Syslog and nothing other, so they required `/dev/log` to works fine.

Thanks for reply.

**Comment by thaJeztah** [comment]
Have you thought of using a (bind-mounted) volume for `/dev/log`?

**Comment by piotrminkina** [comment]
Yes, I thought about it. I don't want bind host `/dev/log` for various reasons (e.g. security and control). I think the best for me is simple install busybox in container and run it in background with parameters like`/bin/busybox syslogd -nSO /proc/1/fd/1`. In this case services that require `/dev/log` now log messages to stdout of main process. Thanks to this solution `docker logs` display logs from container (of course if I run container with default log engine).

Additionally I have full possibilities of Syslog (like facalities and log levels, and I can also forward messages to external host if I want), but with Docker 1.8 I must run container without `--read-only` parameter.

Maybe Docker can create own syslog socket inside container `/dev/log` and depending on selected log engine can forward messages to stdout, stderr and real logging daemon?

**Comment by thaJeztah** [comment]
> I don't want bind host /dev/log for various reason

If the container itself creates a socket in that location, you could use any location as a volume (or just `-v /dev/log` to not *bind-mount* a container)

Something like this may be useful information; https://jpetazzo.github.io/2014/08/24/syslog-docker/

**Comment by piotrminkina** [comment]
Interesting solution, but it have some drawbacks :) If I want one syslog per container then I must create one syslog container per service container. I think this solution is harder to maintain. Thanks to one syslog per service container I can configure syslog with my own standard syslog tag naming, e.g. `container/postfix`, `container/postgresql` etc. and I can do this with `--log-opt syslog-tag` parameter.

From the other side read only `/dev` has also other problems. Please try run

```
docker run --rm -v /tmp/postgresql:/var/lib/postgresql/data -v /run/postgresql -v /tmp postgres:9.4
```

next interrupt by `Ctrl + C` and next rerun with `--read-only` option, as below

```
docker run --rm --read-only -v /tmp/postgresql:/var/lib/postgresql/data -v /run/postgresql -v /tmp postgres:9.4
```

then You should receive message `FATAL:  could not open shared memory segment "/PostgreSQL.1804289383": Read-only file system`. If You check option `dynamic_shared_memory_type` in file `/tmp/postgresql/postgresql.conf` then You see this option has value `posix`. It looks like POSIX Shared Memory not working properly with read only `/dev`.

I think the best solution is that Docker should allow mount `/dev` with read/write even if I run container with `--read-only`. Maybe `--write-dev` parameter is suitable?

**Comment by thaJeztah** [comment]
ping @ewindisch you introduced this change in https://github.com/docker/docker/pull/13669, any thoughts on this?

**Comment by piotrminkina** [comment]
Any decisions?

**Comment by qzio** [comment]
I tried to follow https://jpetazzo.github.io/2014/08/24/syslog-docker/ but it seems as if it doesn't work with newer docker?

```

**Comment by thaJeztah** [comment]
This is fixed by https://github.com/moby/moby/pull/35344 (included in Docker 17.12 and up);

On versions before 17.12:

```bash
$ docker run --rm --read-only busybox sh -c 'touch /dev/log'
touch: /dev/log: Read-only file system
```

On Docker 17.12 and up:

```bash
$ docker run --rm --read-only busybox sh -c 'touch /dev/log'
# (no error)
```
